contracts/cw-names-registry/src/tests.rs

```diff
@@ -0,0 +1,477 @@
+                payment_token_address: token_addr.to_string(),
+                payment_amount,
+            },
+            &[],
+            "DAO Names Registry",
+            None,
+        )
+        .unwrap();
+
+    (names_addr, token_addr)
+}
+
+#[test]
+fn test_instantiate() {
+    let mut app = App::default();
+    let (_names, _token) = setup_test_case(&mut app, Uint128::new(50));
+}
+
+fn register(
+    app: &mut App,
+    names_addr: Addr,
+    amount: Uint128,
+    name: String,
+    sender: Addr,
+    token_addr: Addr,
+) -> AnyResult<AppResponse> {
+    let msg = cw20::Cw20ExecuteMsg::Send {
+        contract: names_addr.to_string(),
+        amount,
+        msg: to_binary(&ReceiveMsg::Register { name }).unwrap(),
+    };
+    app.execute_contract(sender, token_addr, &msg, &[])
+}
+
+fn revoke(app: &mut App, names_addr: Addr, name: String, sender: Addr) -> AnyResult<AppResponse> {
+    let msg = ExecuteMsg::Revoke { name };
+    app.execute_contract(sender, names_addr, &msg, &[])
+}
+
+fn reserve(app: &mut App, names_addr: Addr, name: String, sender: Addr) -> AnyResult<AppResponse> {
+    let msg = ExecuteMsg::Reserve { name };
+    app.execute_contract(sender, names_addr, &msg, &[])
+}
+
+fn transfer_reservation(
+    app: &mut App,
+    names_addr: Addr,
+    name: String,
+    dao: String,
+    sender: Addr,
+) -> AnyResult<AppResponse> {
+    let msg = ExecuteMsg::TransferReservation { name, dao };
+    app.execute_contract(sender, names_addr, &msg, &[])
+}
+
+fn update_config(
+    app: &mut App,
+    names_addr: Addr,
+    new_payment_token_address: Option<String>,
+    new_admin: Option<String>,
+    new_payment_amount: Option<Uint128>,
+    sender: Addr,
+) -> AnyResult<AppResponse> {
+    let msg = ExecuteMsg::UpdateConfig {
+        new_payment_token_address,
+        new_admin,
+        new_payment_amount,
+    };
+    app.execute_contract(sender, names_addr, &msg, &[])
+}
+
+fn query_name(app: &mut App, names_addr: Addr, name: String) -> LookUpNameResponse {
+    let msg = QueryMsg::LookUpName { name };
+    app.wrap().query_wasm_smart(names_addr, &msg).unwrap()
+}
+
+fn query_dao(app: &mut App, names_addr: Addr, dao: String) -> LookUpDaoResponse {
+    let msg = QueryMsg::LookUpDao { dao };
+    app.wrap().query_wasm_smart(names_addr, &msg).unwrap()
+}
+
+fn query_config(app: &mut App, names_addr: Addr) -> Config {
+    let msg = QueryMsg::Config {};
+    app.wrap().query_wasm_smart(names_addr, &msg).unwrap()
+}
+
+#[test]
+fn test_register() {
+    let mut app = App::default();
+    let (names, token) = setup_test_case(&mut app, Uint128::new(50));
+    let other_token = create_token(&mut app); // To be used when sending wrong token
+    let name: &str = "Name";
+
+    // Send wrong token
+    register(
+        &mut app,
+        names.clone(),
+        Uint128::new(50),
+        name.to_string(),
+        Addr::unchecked(DAO_ADDR),
+        other_token,
+    )
+    .unwrap_err();
+
+    // Send too little
```

Comment:
What happens when a DAO sends too much? What should happen when a DAO sends too much? I say: refund the excess. But perhaps that would open up some weird attack or waste gas fees via conditional behavior.